### PR TITLE
sh - implement tags for images

### LINKED ISSUE #6: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by seanhelvey**
tags can now be added through form and displayed on show using specified gem.

Closes #6

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -9,6 +9,15 @@ def test_create__succeed
     assert_redirected_to image_path(Image.last)
   end
 
+  def test_create__with_tags
+    assert_difference('Image.count') do
+      post images_url, params: { image: { url: 'https://www.w3schools.com/w3css/img_lights.jpg', tag_list: 'awesome' } }
```
Comment (grahamotte) [inline test/controllers/images_controller_test.rb]: Probably worth having a test with multiple tags as well

File: db/schema.rb [hunk 2]
```diff
@@ -10,12 +10,37 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2019_04_02_210500) do
+ActiveRecord::Schema.define(version: 2019_04_04_194213) do
```
Comment (grahamotte) [inline db/schema.rb]: Doesn't really matter for this project, but just heads up - we usually don't check in the schema

File: db/migrate/20190404194208_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    create_table :tags do |t|
+      t.string :name
+    end
+
+    create_table :taggings do |t|
+      t.references :tag
+
+      # You should make sure that the column created is
+      # long enough to store the required class names.
+      t.references :taggable, polymorphic: true
```
Comment (grahamotte) [inline db/migrate/20190404194208_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Consider adding a migration on top of these which makes the columns you care about `null: false`

File: test/models/image_test.rb [hunk 4]
```diff
@@ -29,4 +29,16 @@ def test_image__recent_in_order
 
     assert_equal [newer_image, older_image], Image.recent.to_a.first(2)
   end
+
+  def test_image__add_remove_tags
+    image = Image.new(url: 'https://www.w3schools.com/w3css/img_lights.jpg')
+    assert image.valid?
+    assert image.tag_list.empty?
+
+    image.tag_list.add('awesome')
```
Comment (grahamotte) [inline test/models/image_test.rb]: Would also be good to test common error cases like a bunch of commas or empty strings

Comment (seanhelvey) [reply test/models/image_test.rb]: Yup! Done.

File: db/schema.rb (hunk 2 above)
Comment (seanhelvey) [reply db/schema.rb]: rm'd and ignored ✔️

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (seanhelvey) [reply test/controllers/images_controller_test.rb]: Thank you! Added.

File: db/migrate/20190404194208_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 3 above)
Comment (seanhelvey) [reply db/migrate/20190404194208_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: We can talk about this one tomorrow 👍

**Comment by seanhelvey** [comment]
Not sure what you mean @grahamotte . I see tags under the image after uploading one on the image show, which is what I thought the requirement was, but maybe you are saying they should be visible on the index as well? If so, will add that no prob.

**Comment by grahamotte** [comment]
Yeah it looks like it shows when you submit, but there's no way (I saw at least) to get back to that page

File: db/schema.rb (hunk 2 above)
Comment (seanhelvey) [reply db/schema.rb]: Hm removing the schema broke CI with this error message "/home/circleci/repo/db/schema.rb doesn't exist yet. Run `rails db:migrate` to create it, then try again. If you do not intend to use a database, you should instead alter /home/circleci/repo/config/application.rb to limit the frameworks that will be loaded." https://circleci.com/gh/AppFolioOnboarding/image-sharer-seanhelvey/34?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by seanhelvey** [comment]
Got it. I added a link from each image on index to its respective show 👍

File: db/schema.rb (hunk 2 above)
Comment (grahamotte) [reply db/schema.rb]: Ah well then

Comment (seanhelvey) [reply db/schema.rb]: Cool added it back
